### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by isaacabraham**
Another über PR. Sorry. This one does the following:

1. Splits out all the ARM resources into their own namespace files. This adds a *lot* of files but it will make it easier to navigate and maintain (and reduce risk of merge conflicts).
2. I've rationalised the naming here of the core abstractions based on feedback:
   * The "top level" abstraction is now simply `IBuilder`.
   * The "low leve" one is now called `IArmResource`.
3. All Builders now live in `Farmer.Builders`, not `Farmer.Resources` (this is a breaking change to clients but a simple one to fix).
4. All IArmResources live in `Farmer.Arm`.
5. Added an override to `add_resource` so that you can add an `IArmResource` directly to the arm { } builder.
6. Made a tiny change to the `IArmResource` interface - instead of having a method `ToArmObject` that returns the raw json-ready object, it's now just a property called `JsonValue`.

### CODE REVIEW

File: src/Farmer/Arm/Sql.fs [hunk 1]
```diff
@@ -0,0 +1,77 @@
+[<AutoOpen>]
+module Farmer.Arm.Sql
+
+open Farmer
+
+type Server =
```
Comment (dburriss) [inline src/Farmer/Arm/Sql.fs:6]: I noticed this with the Postgres implementation but with AutoOpen and using a generic name like `Server` we are going to run into ambiguity issues. At some stage soon we may want to change this to `SqlServer`. Doesn't have to be this PR though.

File: src/Farmer/Types.fs [hunk 2]
```diff
@@ -13,10 +13,10 @@ type ResourceName =
         | r -> r
     member this.Map mapper = match this with ResourceName r -> ResourceName (mapper r)
 
-/// An Azure ARM resource
-type IResource =
+/// An Azure ARM resource value, which can be serialized to JSON.
+type IArmResource =
     abstract member ResourceName : ResourceName
-    abstract member ToArmObject : unit -> obj
+    abstract member JsonValue : obj
```
Comment (dburriss) [inline src/Farmer/Types.fs:19]: I am being pedantic here, so really I am fine with this name :)
`JsonValue` makes me expect a Json string. Sure then why is the type not string?

What about `JsonObject` or `JsonModel`.

Just some suggestions and I think as it is also fine.

File: src/Farmer/Arm/ContainerRegistry.fs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+[<AutoOpen>]
+module Farmer.Arm.ContainerRegistry
+
+open Farmer
+
+type Registries =
```
Comment (dburriss) [inline src/Farmer/Arm/ContainerRegistry.fs:6]: `Registry` maybe?

File: src/Farmer/Types.fs (hunk 2 above)
Comment (isaacabraham) [reply src/Farmer/Types.fs:19]: Either of those work for me - I'll update.

File: src/Farmer/Arm/Sql.fs (hunk 1 above)
Comment (isaacabraham) [reply src/Farmer/Arm/Sql.fs:6]: Mmmm. Perhaps we should turn AutoOpen off for the Arm ones. Not sure. I wanted to keep the names identical to the official names so it aligned with the ARM terminology (including namespaces). Perhaps we will need to rethink this.

File: src/Farmer/Arm/ContainerRegistry.fs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    member this.LoginServer =
+        (sprintf "reference(resourceId('Microsoft.ContainerRegistry/registries', '%s'),'2019-05-01').loginServer" this.Name.Value)
+        |> ArmExpression
+    interface IArmResource with
+        member this.ResourceName = this.Name
+        member this.JsonValue =
+            {| name = this.Name.Value
+               ``type`` = "Microsoft.ContainerRegistry/registries"
+               apiVersion = "2019-05-01"
+               sku = {| name = this.Sku.ToString().Replace("_", ".") |}
```
[the remark this answers is not in the document]
Comment (isaacabraham) [reply src/Farmer/Arm/ContainerRegistry.fs:20]: Funny you mention that - I just noticed this myself. Yes, it is. There another branch I'm working on which accidentally snuck in here (caused by the gitignore file ignoring the ARM folder!). Happy to remove this although it's harmless, and will be overwritten by the other branch if it goes in.

File: src/Farmer/Arm/ContainerRegistry.fs (hunk 3 above)
Comment (isaacabraham) [reply src/Farmer/Arm/ContainerRegistry.fs:6]: Again, this is based on the official ARM namespace.

File: src/Farmer/Arm/ContainerRegistry.fs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dburriss) [reply src/Farmer/Arm/ContainerRegistry.fs:20]: Yeah, it's harmless. Just noticed it so I thought I would point it out. No rush.

**Pull request merged**